msticpy/nbtools/timeline.py

```diff
@@ -0,0 +1,940 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+"""Module for common display functions."""
+from typing import Any, Union, Set, Dict
+
+import pandas as pd
+from bokeh.io import output_notebook, show
+from bokeh.models import (
+    ColumnDataSource,
+    DatetimeTickFormatter,
+    HoverTool,
+    Label,
+    Legend,
+    RangeTool,
+    Title,
+)
+
+# pylint: disable=no-name-in-module
+from bokeh.palettes import viridis
+
+# pylint: enable=no-name-in-module
+from bokeh.plotting import figure, reset_output
+from bokeh.layouts import column
+
+from .._version import VERSION
+from .utility import export
+
+__version__ = VERSION
+__author__ = "Ian Hellen"
+
+
+# Constants
+_WRAP = 50
+_WRAP_CMDL = "WrapCmdl"
+
+
+@export
+def display_timeline(
+    data: Union[pd.DataFrame, dict],
+    time_column: str = "TimeGenerated",
+    source_columns: list = None,
+    **kwargs,
+) -> figure:
+    """
+    Display a timeline of events.
+
+    Parameters
+    ----------
+    data : Union[dict, pd.DataFrame]
+        Either
+        dict of data sets to plot on the timeline with the following structure::
+
+            Key (str) - Name of data set to be displayed in legend
+            Value (Dict[str, Any]) - containing:
+                data (pd.DataFrame) - Data to plot
+                time_column (str, optional) - Name of the timestamp column
+                source_columns (list[str], optional) - source columns to use
+                    in tooltips
+                color (str, optional) - color of datapoints for this data
+            If any of the last values are omitted, they default to the values
+            supplied as parameters to the function (see below)
+
+        Or
+        DataFrame as a single data set or grouped into individual
+        plot series using the `group_by` parameter
+    time_column : str, optional
+        Name of the timestamp column
+        (the default is 'TimeGenerated')
+    source_columns : list, optional
+        List of default source columns to use in tooltips
+        (the default is None)
+
+    Other Parameters
+    ----------------
+    title : str, optional
+        Title to display (the default is None)
+    alert : SecurityAlert, optional
+        Add a reference line/label using the alert time (the default is None)
+    ref_event : Any, optional
+        Add a reference line/label using the alert time (the default is None)
+    ref_time : datetime, optional
+        Add a reference line/label using `ref_time` (the default is None)
+    group_by : str
+        (where `data` is a DataFrame)
+        The column to group timelines on
+    sort_by : str
+        (where `data` is a DataFrame)
+        The column to order timelines on
+    legend: str, optional
```

Comment:
I think the legend is pretty key to comprehension of the timeline, I would suggest we default the legend to being inline and provide the option to remove/move it if wanted.

Reply:
Legend on (inline, centered vertically) by default unless single data series with no legend column